### PR TITLE
Feature/0464 refactor seqview

### PR DESCRIPTION
**Pull request by seantalts**
#### Submisison Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Refactor most common use case of VectorView (taking either a primitive or a sequence of primitives and treating it seamlessly as a sequence of primitives) into scalar_seq_view, which is much simpler. 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Dual Space LLC

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
Is that `[WIP]` supposed to mean this is work in progress?  That makes sense if you are just using the test harness.  But you'll need to let someone know when you want it reviewed.

**Comment by bob-carpenter** [comment]
OK --- we can change that to a label at some point if we
get too many of these and need to sort.  Let's hope it doesn't
come to that.  Daniel's had some WIPs for a whole as issues.
Usually we kill them after a while pending completion 
so our pull requests aren't cluttered.

- Bob

**Comment by syclik** [comment]
Yeah -- we shouldn't have too many of those.

The last pull request went from what I thought was complete to realizing it
wasn't; we can kill it and I can reopen later if that helps.

**Comment by seantalts** [comment]
@bob-carpenter Okay, tests pass and I'm marking this as non-WIP. I think it might be a good place to start with the refactor, keeping in mind your advice to keep pull requests as small as possible. Bob, since you and I paired on the design together do you think we should get a 3rd party to review it? And then whoever wants to can double check that I did the substitutions correctly... Actually maybe I should have just put up the design first and then done the replacements in a 2nd PR? What do you guys think?

File: stan/math/prim/scal/meta/scalar_seq_view.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#ifndef STAN_MATH_PRIM_SCAL_META_SCALAR_SEQ_VIEW_HPP
+#define STAN_MATH_PRIM_SCAL_META_SCALAR_SEQ_VIEW_HPP
+
+#include <stan/math/prim/scal/meta/scalar_type.hpp>
+
+namespace stan {
+
+  template <typename C, typename T = typename scalar_type<C>::type>
```
Comment (syclik) [inline stan/math/prim/scal/meta/scalar_seq_view.hpp]: Add documentation for the class. See https://github.com/stan-dev/stan/wiki/How-to-Write-Doxygen-Doc-Comments

File: stan/math/prim/scal/meta/scalar_seq_view.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    int size() const {
+      return c_.size();
+    }
+
+  private:
+    const C& c_;
+  };
+
+  template <typename T>
```
Comment (syclik) [inline stan/math/prim/scal/meta/scalar_seq_view.hpp]: Add doc describing what this specialization is.

File: stan/math/prim/mat/meta/scalar_seq_view.hpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#ifndef STAN_MATH_PRIM_MAT_META_SCALAR_SEQ_VIEW_HPP
+#define STAN_MATH_PRIM_MAT_META_SCALAR_SEQ_VIEW_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/scal/meta/scalar_seq_view.hpp>
+
+namespace stan {
+  template <typename T>
```
Comment (syclik) [inline stan/math/prim/mat/meta/scalar_seq_view.hpp]: Add doc about the template specialization.

File: test/unit/math/prim/mat/meta/scalar_seq_view_test.cpp [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  EXPECT_FLOAT_EQ(v(0), sv[0]);
+  EXPECT_FLOAT_EQ(v(1), sv[1]);
+  EXPECT_FLOAT_EQ(v(2), sv[2]);
+  EXPECT_FLOAT_EQ(v(3), sv[3]);
+
+  EXPECT_EQ(v.size(), sv.size());
+}
+
+TEST(MetaTraits, ScalarSeqViewMatrix) {
+  expect_scalar_seq_view_values(Eigen::VectorXd(4));
```
Comment (syclik) [inline test/unit/math/prim/mat/meta/scalar_seq_view_test.cpp]: if you're bored, you can look at scoped traces in Google Test to give context when this fails. No need to take action, but it is a useful feature of Google Test.

File: test/unit/math/prim/mat/meta/scalar_type_test.cpp [hunk 5]
```diff
@@ -1,17 +1,21 @@
...
-  EXPECT_FLOAT_EQ(1, x);
+  expect_same_type<double, scalar_type<MatrixXd>::type>();
+  expect_same_type<double, scalar_type<VectorXd>::type>();
+  expect_same_type<double, scalar_type<RowVectorXd>::type>();
 
-  // hack to get value of template into Google test macro 
-  bool b3 = is_same<double, scalar_type<Matrix<double, Dynamic, Dynamic> >::type>::value;
-  EXPECT_TRUE(b3);
+  // TODO(Sean): This behavior doesn't seem to make sense prima facie
+  // Likely change this so that we would expect `double` here instead.
```
Comment (syclik) [inline test/unit/math/prim/mat/meta/scalar_type_test.cpp]: Just commenting; I think `scalar_type` just unwraps one layer. I think there's a recursive version kicking around. But either way, this is improperly named.

Comment (seantalts) [inline test/unit/math/prim/mat/meta/scalar_type_test.cpp]: It appears to be recursive on `std::vector` but not `Matrix` 🌝

File: test/unit/math/prim/mat/meta/scalar_seq_view_test.cpp (hunk 4 above)
Comment (seantalts) [inline test/unit/math/prim/mat/meta/scalar_seq_view_test.cpp]: Ooh, sounds cool! Will check it out, thanks.

File: stan/math/prim/scal/meta/scalar_seq_view.hpp [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  class scalar_seq_view {
+  public:
+    explicit scalar_seq_view(const C& c) : c_(c) {}
+
+    /**
+     * Segfaults if out of bounds.
+     * @param i index
+     * @return the element at the specified position in the container
+     */
+    const T& operator[](int i) const {
```
Comment (seantalts) [inline stan/math/prim/scal/meta/scalar_seq_view.hpp]: What do you guys think about these `operator[]`s being inlined? I'm still getting a feel for where that's appropriate.

**Comment by seantalts** [comment]
jenkins, please retest this

**Comment by seantalts** [comment]
jenkins, retest this please

**Comment by bob-carpenter** [comment]
We're not that worried about size.  Yes, they should probably be inline.  There's two aspects of inline---the compiler hint and the translation unit duplication feature.  Member functions automatically get the latter.  See this:

http://stackoverflow.com/questions/26241023/different-implementations-of-inline-functions-in-different-translation-units

**Comment by seantalts** [comment]
Re: inline - I just looked into it a little more, and some stuff suggests that all member functions declared inside a class are implicitly inline and that marking it as inline does nothing? http://stackoverflow.com/questions/9370493/inline-function-members-inside-a-class

**Comment by betanalpha** [comment]
Correct, that’s what I was trying to yell in the background.

**Comment by seantalts** [comment]
Schwoops. Well, I'll leave them unmarked then.

**Comment by bob-carpenter** [comment]
Yes, member functions are automatically inline in the sense that you
can use the definition in multiple translation units.

But what I was trying to say is that I'm not sure if that implies
they are also inline in the nudge the compiler into unfolding sense.

- Bob

**Comment by seantalts** [comment]
Bob, gotcha. I think from what I was reading on that SO link that literally nothing changes in the compiler if you mark a member function inline, though it's hard to imagine that being perfectly consistent across all compilers...

Dan, I added doc if you want to take a quick look whenever you have a chance and see if it looks good now? Thanks!

**Comment by syclik** [comment]
Looks good to me. Just approved.

**Pull request merged**